### PR TITLE
libblkid: xfs: add checksum support

### PR DESCRIPTION
**Pull request by t-8ch**

### CODE REVIEW

**Comment by mator** [comment]
getting the same error on this branch as in #1813 
```
mator@ttip:/1/mator/util-linux/tests$ ./run.sh --show-diff blkid/low-probe

-------------------- util-linux regression tests --------------------

                    For development purpose only.
                 Don't execute on production system!

       kernel: 5.19.0-1-sparc64-smp

      options: --show-diff \
               --srcdir=/1/mator/util-linux/tests/.. \
               --builddir=/1/mator/util-linux/tests/..

        blkid: superblocks probing            ...
                : adaptec-raid                ... OK
                : bcache-B                    ... OK
...
                : vmfs_volume                 ... OK
                : xfs-log                     ... OK
                : xfs-v5                      ...
diff-{{{
--- /1/mator/util-linux/tests/expected/blkid/low-probe-xfs-v5   2022-05-30 16:25:21.647438447 +0300
+++ /1/mator/util-linux/tests/output/blkid/low-probe-xfs-v5     2022-09-16 21:14:07.621659382 +0300
@@ -1,10 +0,0 @@
-ID_FS_BLOCK_SIZE=512
-ID_FS_FSBLOCKSIZE=4096
-ID_FS_FSLASTBLOCK=5120
-ID_FS_FSSIZE=17469440
-ID_FS_LABEL=test-xfs-v5
-ID_FS_LABEL_ENC=test-xfs-v5
-ID_FS_TYPE=xfs
-ID_FS_USAGE=filesystem
-ID_FS_UUID=3fdcb214-0f39-466d-a81a-a1fb114fe7cd
-ID_FS_UUID_ENC=3fdcb214-0f39-466d-a81a-a1fb114fe7cd
}}}-diff

 FAILED (blkid/low-probe-xfs-v5)
                : xfs                         ... OK
                : zfs                         ... OK
                : zonefs                      ... OK
           ... FAILED (2 from 116 sub-tests)

---------------------------------------------------------------------
  1 tests of 1 FAILED

      blkid/low-probe
---------------------------------------------------------------------
mator@ttip:/1/mator/util-linux/tests$
```

but works with system installed `libblkid.so.1`:
```
mator@ttip:/1/mator/util-linux$ xzcat tests/ts/blkid/images-fs/xfs-v5.img.xz > xfs-v5.img
mator@ttip:/1/mator/util-linux$ .libs/blkid -p -o udev xfs-v5.img
ID_FS_LABEL=test-xfs-v5
ID_FS_LABEL_ENC=test-xfs-v5
ID_FS_UUID=3fdcb214-0f39-466d-a81a-a1fb114fe7cd
ID_FS_UUID_ENC=3fdcb214-0f39-466d-a81a-a1fb114fe7cd
ID_FS_BLOCK_SIZE=512
ID_FS_TYPE=xfs
ID_FS_USAGE=filesystem
```

**Comment by t-8ch** [comment]
@mator Thanks for the report. It show now also for on big-endian (validated on s390x).

**Comment by mator** [comment]
thanks, works now (tested on sparc64)

File: libblkid/src/superblocks/xfs.c [hunk 1]
```diff
@@ -118,6 +141,29 @@ static void sb_from_disk(struct xfs_super_block *from,
...
+	to->sb_bad_features2 = be32_to_cpu(from->sb_bad_features2);
+	to->sb_features_compat = be32_to_cpu(from->sb_features_compat);
+	to->sb_features_ro_compat = be32_to_cpu(from->sb_features_ro_compat);
+	to->sb_features_incompat = be32_to_cpu(from->sb_features_incompat);
+	to->sb_features_log_incompat = be32_to_cpu(from->sb_features_log_incompat);
+	to->sb_crc = be32_to_cpu(from->sb_crc);
+	to->sb_spino_align = be32_to_cpu(from->sb_spino_align);
+	to->sb_pquotino = be64_to_cpu(from->sb_pquotino);
+	to->sb_lsn = be64_to_cpu(from->sb_lsn);
+	to->sb_rrmapino = be64_to_cpu(from->sb_rrmapino);
```
Comment (karelzak) [inline libblkid/src/superblocks/xfs.c:170]: I guess the new v5 superblock members should be read by sb_from_disk() only when `if ((sbp->sb_versionnum & 0x0f) == 5)`, right?

Comment (t-8ch) [reply libblkid/src/superblocks/xfs.c:170]: In theory yes. But as the memory will be there anyways it should be fine to read it unconditionally.

File: libblkid/src/superblocks/xfs.c [hunk 2]
```diff
@@ -199,7 +204,20 @@ static int xfs_verify_sb(struct xfs_super_block *ondisk)
 	    sbp->sb_dblocks < XFS_MIN_DBLOCKS(sbp))
 		return 0;
 
-	/* TODO: version 5 has also checksum CRC32, maybe we can check it too */
+	if ((sbp->sb_versionnum & 0x0f) == 5) {
+		if (!(sbp->sb_versionnum | XFS_SB_VERSION_MOREBITSBIT))
+			return 0;
+		if (!(sbp->sb_features2 | XFS_SB_VERSION2_CRCBIT))
+			return 0;
```
Comment (karelzak) [inline libblkid/src/superblocks/xfs.c:211]: Coverity scan reports this as an issue:
```
207         if ((sbp->sb_versionnum & 0x0f) == 5) {
>>>     CID 381139:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
>>>     "sbp->sb_versionnum | 0x8000" is always 1/true regardless of the values of its operand. This occurs as the logical operand of "!".
208             if (!(sbp->sb_versionnum | XFS_SB_VERSION_MOREBITSBIT))
209                 return 0;
210             if (!(sbp->sb_features2 | XFS_SB_VERSION2_CRCBIT))
211                 return 0;
```

Is the '|' operator expected there? At first glance, I'd expect '&'.

**Comment by t-8ch** [comment]
Yes, this looks wrong indeed.

Do you want me to send a fix?

**Comment by karelzak** [comment]
Fixed.

**Pull request merged**